### PR TITLE
pwncat: initial commit

### PR DESCRIPTION
**Pull request by cytopia**
### `pwncat`: initial commit

> pwncat is a sophisticated bind and reverse shell handler with many features as well as a drop-in replacement or compatible complement to netcat, ncat or socat.

* https://github.com/cytopia/pwncat
* http://pwncat.org/

### CODE REVIEW

File: packages/pwncat/PKGBUILD [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# This file is part of BlackArch Linux ( https://www.blackarch.org/ ).
+# See COPYING for license details.
+
+pkgname=pwncat
+pkgver=0.0.23-alpha
```
Comment (noraj) [inline packages/pwncat/PKGBUILD:5]:
```
$ ba-dev -b
Syncing chroot copy /home/noraj/blackarch_chroot/blackarch/ with /home/noraj/blackarch_chroot/root/ ...
[sudo] password for noraj: 
Chroot environment is ready!
Consideer installing pkgcheck with pip install pkgcheck-arch --user for automatic PKGBUILD syntax error checking.
==> ERROR: pkgver is not allowed to contain colons, forward slashes, hyphens or whitespace.
==> ERROR: Could not download sources.
Failed to build the package.
```

File: packages/pwncat/PKGBUILD [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+pkgrel=1
+groups=('blackarch' 'blackarch-scanner' 'blackarch-backdoor' 'blackarch-networking' 'blackarch-proxy')
+pkgdesc='Bind and reverse shell handler with FW/IDS/IPS evasion, self-inject and port-scanning'
+url='http://pwncat.org/'
+license=('MIT')
+arch=('any')
+depends=('python')
+source=("$pkgname-$pkgver.tar.gz::https://github.com/cytopia/pwncat/archive/v$pkgver.tar.gz")
```
Comment (noraj) [inline packages/pwncat/PKGBUILD:13]:
```
$ ba-dev -b
Syncing chroot copy /home/noraj/blackarch_chroot/blackarch/ with /home/noraj/blackarch_chroot/root/ ...
[sudo] password for noraj: 
Chroot environment is ready!
Consideer installing pkgcheck with pip install pkgcheck-arch --user for automatic PKGBUILD syntax error checking.
==> ERROR: pkgver is not allowed to contain colons, forward slashes, hyphens or whitespace.
==> ERROR: Could not download sources.
Failed to build the package.
```

**Comment by noraj** [comment]
Thanks for the PR, can you fix the PKGBUILD and test it with https://github.com/Edu4rdSHL/blackarch-devtools ?

**Comment by cytopia** [comment]
I was not able to get blackarch-devtools to work, but tested it via the following instead:

```bash
makepkg
makepkg --install
```

File: packages/pwncat/PKGBUILD (hunk 1 above)
Comment (cytopia) [reply packages/pwncat/PKGBUILD]: Fixed

File: packages/pwncat/PKGBUILD (hunk 2 above)
Comment (cytopia) [reply packages/pwncat/PKGBUILD]: Fixed

**Comment by Edu4rdSHL** [comment]
Thank you @cytopia. What's the error with devtools? You just need to download the binary (or compile it), had the BlackArch repos in your host system and execute `ba-dev -s` to setup the chroot environment, after that just use `ba-dev -b` and `ba-dev -p path_to_package.pkg.tar.xz -e command_to_execute (usually the name of the package binary)`.

**Comment by cytopia** [comment]
@Edu4rdSHL I was trying to automate this with Docker and ran into the following error:

```bash
[arch@24762ff4ff39 data]$ id
uid=1000(arch) gid=1000(arch) groups=1000(arch),998(wheel)

[arch@24762ff4ff39 data]$ ls -lap
total 4616
drwxr-xr-x 4 arch arch    4096 Jun 22 10:51 ./
drwxr-xr-x 1 root root    4096 Jun 22 10:49 ../
drwxr-xr-x 3 arch arch    4096 Jun 22 10:51 pkg/
-rw-r--r-- 1 arch arch    1054 Jun 22 10:46 PKGBUILD
-rw-r--r-- 1 arch arch   65788 Jun 22 10:51 pwncat-0.1.0-1-any.pkg.tar.xz
-rw-r--r-- 1 arch arch 4630507 Jun 22 10:51 pwncat-0.1.0.tar.gz
drwxr-xr-x 3 arch arch    4096 Jun 22 10:51 src/

[arch@24762ff4ff39 data]$ ba-dev -b
Syncing chroot copy /home/arch/blackarch_chroot/blackarch/ with /home/arch/blackarch_chroot/root/ ...
Chroot environment doesn't exists. Please use the -s option first.
==> ERROR: No chroot dir defined, or invalid path '/home/arch/blackarch_chroot'
Failed to build the package.

[arch@24762ff4ff39 data]$ ba-dev -s
Creating chroot directory with name: /home/arch/blackarch_chroot
thread 'main' panicked at 'An error as ocurred while creating the chroot directoy.: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/libcore/result.rs:1165:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

**Comment by Edu4rdSHL** [comment]
Hello @, I think that the user home directory doesn't exist. Can you check if `/home/arch` exists? Thanks.

**Comment by cytopia** [comment]
That was it, as well the docker container needs to run in privileged mode as well needs the host's `/proc/` mounted.
It's not perfect as it still errors out for the chroot env, but you can see that it at least passes:
```
Failed to retrieve machine ID: No such file or directory
Attempted to remove disk file system under "/run/systemd/nspawn/propagate/root", and we can't allow that.
Failed to install base packages into chroot environment.
Syncing chroot copy /home/arch/blackarch_chroot/blackarch/ with /home/arch/blackarch_chroot/root/ ...
Chroot environment is ready!
==> Making package: pwncat 0.1.0-1 (Mon Jun 22 13:36:16 2020)
==> Retrieving sources...
  -> Found pwncat-0.1.0.tar.gz
==> Validating source files with sha512sums...
    pwncat-0.1.0.tar.gz ... Passed
Failed to retrieve machine ID: No such file or directory
Attempted to remove disk file system under "/run/systemd/nspawn/propagate/blackarch", and we can't allow that.
==> ERROR: Build failed, check /home/arch/blackarch_chroot/blackarch/build
Failed to build the package.
```

The whole thing automated is here:
```bash
#!/usr/bin/env bash
cat << EOF > "${DESTPATH}/start.sh"
#!/bin/sh
chmod 0777 /data
pacman -Sy
yes | pacman -S fakeroot binutils python sudo devtools wget gettext
wget https://github.com/Edu4rdSHL/blackarch-devtools/releases/latest/download/blackarch-devtools
chmod +x blackarch-devtools
cp blackarch-devtools /usr/bin/ba-dev
useradd arch
usermod -aG wheel arch
mkdir /home/arch
chown arch:arch /home/arch
echo "%wheel ALL=(ALL) NOPASSWD: ALL" >> /etc/sudoers
echo "root ALL=(ALL) NOPASSWD: ALL" >> /etc/sudoers
# Test build/install
su - arch -c 'cd /data && makepkg'
su - arch -c 'cd /data && yes | makepkg --install'
# Test ba
su - arch -c 'cd /data && ba-dev -s' || true
su - arch -c 'cd /data && ba-dev -b' || true
EOF
chmod +x "${DESTPATH}/start.sh"

docker run \
	--rm \
	-w /data \
	-v "${DESTPATH}/PKGBUILD:/data/PKGBUILD" \
	-v "${DESTPATH}/start.sh:/start.sh" \
	-v /proc:/proc \
	--privileged \
	--entrypoint=/start.sh \
	archlinux
```

On the other hand, only doing:
```
su - arch -c 'cd /data && makepkg'
su - arch -c 'cd /data && yes | makepkg --install'
```
are also perfectly fine, as they complaint on any issue (e.g.: wrong shasum, etc)

**Comment by Edu4rdSHL** [comment]
I see, try using it outside of docker, there aren't data loss risks as everything is done in a chroot environment.

**Comment by cytopia** [comment]
> I see, try using it outside of docker

For manual checks most likely, but then for automation in CI I require Docker as I cannot ensure what the underlying OS is and provides (e.g.: Arch-based)

**Pull request merged**